ProbMask/runs/mnist_conv_fashion_mnist-pr0.1/probmask/prune_rate=0.1/settings.txt
Namespace(K=2, T=1, TA=True, arch='mnist_conv', batch_size=1000, bn_type='LearnedBatchNorm', config='configs/mnist_conv_fashion_mnist-pr0.1.yaml', conv_type='ProbMaskConv', data='dataset/', epochs=100, evaluate=False, finetune=True, finetuning=False, first_layer_dense=False, first_layer_type=None, freeze_weights=False, gpu=0, init='kaiming_normal', init_weights='', iterative=True, label_smoothing=None, last_layer_dense=False, log_dir=None, low_data=1, lr=0.006, lr_adjust=30, lr_gamma=0.1, lr_policy='cosine_lr', mode='fan_in', momentum=0.9, multigpu=[0], name='probmask', nesterov=False, no_bn_decay=False, nonlinearity='relu', num_classes=10, one_batch=False, optimizer='adam', pr_start=1.0, pretrained=None, print_freq=10, prune_rate=0.1, random_subnet=False, rep_count='/', resume='', resume_train_weights='', sample_from_training_set=True, save_every=-1, scale_fan=False, score_init_constant=1, seed=None, set='FashionMNIST', start_epoch=None, te=0.6, train_weights_at_the_same_time=True, trainer='default', ts=0.16, use_running_stats=False, warmup_length=0, weight_decay=0, weight_opt_lr=0.1, width_mult=1.0, workers=4)

ProbMask/runs/mnist_conv_fashion_mnist-pr0.1/probmask/prune_rate=0.9/settings.txt
Namespace(K=2, T=1, TA=True, arch='mnist_conv', batch_size=1000, bn_type='LearnedBatchNorm', config='configs/mnist_conv_fashion_mnist-pr0.1.yaml', conv_type='ProbMaskConv', data='dataset/', epochs=30, evaluate=False, finetune=False, finetuning=False, first_layer_dense=False, first_layer_type=None, freeze_weights=False, gpu=0, init='kaiming_normal', init_weights='', iterative=True, label_smoothing=None, last_layer_dense=False, log_dir=None, low_data=1, lr=0.006, lr_adjust=30, lr_gamma=0.1, lr_policy='cosine_lr', mode='fan_in', momentum=0.9, multigpu=[0], name='probmask', nesterov=False, no_bn_decay=False, nonlinearity='relu', num_classes=10, one_batch=False, optimizer='adam', pr_start=1.0, pretrained=None, print_freq=10, prune_rate=0.9, random_subnet=False, rep_count='/', resume='', resume_train_weights='', sample_from_training_set=True, save_every=-1, scale_fan=False, score_init_constant=1, seed=None, set='FashionMNIST', start_epoch=None, te=0.6, train_weights_at_the_same_time=True, trainer='default', ts=0.16, use_running_stats=False, warmup_length=0, weight_decay=0, weight_opt_lr=0.1, width_mult=1.0, workers=4)

ProbMask/args.py
import argparse
import sys
import yaml

from configs import parser as _parser

args = None


def parse_arguments():
    parser = argparse.ArgumentParser(description="PyTorch ImageNet Training")

    # General Config
    parser.add_argument(
        "--data", help="path to dataset base directory", default="dataset/"
    )
    parser.add_argument("--optimizer", help="Which optimizer to use", default="sgd")
    parser.add_argument("--set", help="name of dataset", type=str, default="ImageNet")
    parser.add_argument(
        "-a", "--arch", metavar="ARCH", default="ResNet18", help="model architecture"
    )
    parser.add_argument(
        "--config", help="Config file to use (see configs dir)", default=None
    )
    parser.add_argument(
        "--log-dir", help="Where to save the runs. If None use ./runs", default=None
    )
    parser.add_argument(
        "-j",
        "--workers",
        default=20,
        type=int,
        metavar="N",
        help="number of data loading workers (default: 20)",
    )
    parser.add_argument(
        "--epochs",
        default=90,
        type=int,
        metavar="N",
        help="number of total epochs to run",
    )
    parser.add_argument(
        "--start-epoch",
        default=None,
        type=int,
        metavar="N",
        help="manual epoch number (useful on restarts)",
    )
    parser.add_argument(
        "-b",
        "--batch-size",
        default=256,
        type=int,
        metavar="N",
        help="mini-batch size (default: 256), this is the total "
             "batch size of all GPUs on the current node when "
             "using Data Parallel or Distributed Data Parallel",
    )
    parser.add_argument(
        "--lr",
        default=6e-3,
        type=float,
        help="initial learning rate",
    )
    parser.add_argument(
        "--warmup_length", default=0, type=int, help="Number of warmup iterations"
    )
    parser.add_argument(
        "--momentum", default=0.9, type=float, metavar="M", help="momentum"
    )
    parser.add_argument(
        "--wd",
        "--weight-decay",
        default=1e-4,
        type=float,
        metavar="W",
        help="weight decay (default: 1e-4)",
        dest="weight_decay",
    )
    parser.add_argument(
        "-p",
        "--print-freq",
        default=10,
        type=int,
        metavar="N",
        help="print frequency (default: 10)",
    )
    parser.add_argument("--num-classes", default=10, type=int)
    parser.add_argument(
        "--resume",
        default="",
        type=str,
        metavar="PATH",
        help="path to latest checkpoint (default: none)",
    )
    parser.add_argument(
        "--resume_train_weights",
        default="",
        type=str,
        metavar="PATH",
        help="path to latest checkpoint (default: none)",
    )
    parser.add_argument(
        "-e",
        "--evaluate",
        dest="evaluate",
        action="store_true",
        help="evaluate model on validation set",
    )
    parser.add_argument(
        "--pretrained",
        dest="pretrained",
        default=None,
        type=str,
        help="use pre-trained model",
    )
    parser.add_argument(
        "--seed", default=None, type=int, help="seed for initializing training. "
    )
    parser.add_argument(
        "--multigpu",
        default=None,
        type=lambda x: [int(a) for a in x.split(",")],
        help="Which GPUs to use for multigpu training",
    )

    # Learning Rate Policy Specific
    parser.add_argument(
        "--lr-policy", default="cosine_lr", help="Policy for the learning rate."
    )
    parser.add_argument(
        "--lr-adjust", default=30, type=int, help="Interval to drop lr"
    )
    parser.add_argument(
        "--lr-gamma", default=0.1, type=int, help="Multistep multiplier"
    )
    parser.add_argument(
        "--name", default=None, type=str, help="Experiment name to append to filepath"
    )
    parser.add_argument(
        "--save-every", default=-1, type=int, help="Save every ___ epochs"
    )
    parser.add_argument(
        "--prune-rate",
        default=0.0,
        help="Amount of pruning to do during sparse training",
        type=float,
    )

    parser.add_argument(
        "--pr_start",
        default=1.0,
        help="Amount of pruning rate for start",
        type=float,
    )

    parser.add_argument(
        "--low-data", default=1, help="Amount of data to use", type=float
    )
    parser.add_argument(
        "--width-mult",
        default=1.0,
        help="How much to vary the width of the network.",
        type=float,
    )
    parser.add_argument(
        "--nesterov",
        default=False,
        action="store_true",
        help="Whether or not to use nesterov for SGD",
    )
    parser.add_argument(
        "--random-subnet",
        action="store_true",
        help="Whether or not to use a random subnet when fine tuning for lottery experiments",
    )
    parser.add_argument(
        "--one-batch",
        action="store_true",
        help="One batch train set for debugging purposes (test overfitting)",
    )
    parser.add_argument(
        "--conv-type", type=str, default=None, help="What kind of sparsity to use"
    )
    parser.add_argument(
        "--freeze-weights",
        action="store_true",
        help="Whether or not to train only subnet (this freezes weights)",
    )

    parser.add_argument("--mode", default="fan_in", help="Weight initialization mode")
    parser.add_argument(
        "--nonlinearity", default="relu", help="Nonlinearity used by initialization"
    )
    parser.add_argument("--bn-type", default=None, help="BatchNorm type")
    parser.add_argument(
        "--init", default="kaiming_normal", help="Weight initialization modifications"
    )
    parser.add_argument(
        "--no-bn-decay", action="store_true", default=False, help="No batchnorm decay"
    )
    parser.add_argument(
        "--scale-fan", action="store_true", default=False, help="scale fan"
    )
    parser.add_argument(
        "--first-layer-dense", action="store_true", help="First layer dense or sparse"
    )
    parser.add_argument(
        "--last-layer-dense", action="store_true", help="Last layer dense or sparse"
    )
    parser.add_argument(
        "--label-smoothing",
        type=float,
        help="Label smoothing to use, default 0.0",
        default=None,
    )
    parser.add_argument(
        "--first-layer-type", type=str, default=None, help="Conv type of first layer"
    )
    parser.add_argument(
        "--trainer", type=str, default="default", help="cs, ss, or standard training"
    )
    parser.add_argument(
        "--score-init-constant",
        type=float,
        default=1,
    )

    parser.add_argument(
        "--K",
        type=int,
        default=2,
        help="Sample K nets",
    )

    parser.add_argument(
        "--T",
        type=float,
        default=1,
        help="Temperature Annealing Parameter",
    )

    parser.add_argument(
        "--TA",
        default=True,
        action="store_true",
        help="Temperature Annealing",
    )

    parser.add_argument(
        "--init_weights",
        type=str,
        default="",
        help="init weights loc",
    )

    parser.add_argument(
        "--use_running_stats",
        default=False,
        action="store_true",
        help="Whether use bn running stats",
    )

    parser.add_argument(
        "--iterative",
        default=True,
        action="store_true",
        help="Whether use iterative pruning",
    )

    parser.add_argument(
        "--train_weights_at_the_same_time",
        default=False,
        action="store_true",
        help="Whether train_weights at the same time",
    )

    parser.add_argument(
        "--sample_from_training_set",
        default=True,
        action="store_true",
        help="Whether sample from training set",
    )

    parser.add_argument(
        "--finetune",
        default=False,
        action="store_true",
        help="Whether finetune",
    )

    parser.add_argument(
        "--weight_opt_lr",
        type=float,
        default=0.1,
        help="lr for weight training at the same time",
    )

    parser.add_argument(
        "--ts",
        type=float,
        default=0.16,
        help="ts",
    )

    parser.add_argument(
        "--te",
        type=float,
        default=0.6,
        help="te",
    )

    args = parser.parse_args()

    # Allow for use from notebook without config file
    if len(sys.argv) > 1:
        get_config(args)

    return args


def get_config(args):
    # get commands from command line
    override_args = _parser.argv_to_vars(sys.argv)

    # load yaml file
    yaml_txt = open(args.config).read()

    # override args
    loaded_yaml = yaml.load(yaml_txt, Loader=yaml.FullLoader)
    for v in override_args:
        loaded_yaml[v] = getattr(args, v)

    print(f"=> Reading YAML config from {args.config}")
    args.__dict__.update(loaded_yaml)


def run_args():
    global args
    if args is None:
        args = parse_arguments()


run_args()
